### PR TITLE
Allow the script to skip tasks from the pipeline

### PR DESCRIPTION
**Pull request by kevinkim9264**
With this PR, the invoker can skip certain tasks of the pipeline. This is useful when the invoker needs to skip a task (due to a known error) or wants to invoke a certain task only by skipping all the other tasks.

The pipeline skips a task by not instantiating the Task object and not adding to the pipeline. The other downstream dependent tasks are not affected because the object just takes the value of the parent task.

e.g. Given that    export_secrets => (depends on) export_groups
                             export_clusters => export_secrets
if export_secrets is skipped, export_clusters => export_groups, because export_secrets = export_groups in this case.

Usage:
```
python3 migration_pipeline.py --profile staging_src --import-pipeline --use-checkpoint --cluster-name kevinTestExport --skip-tasks instance_profiles users groups workspace_item_log workspace_acls notebooks secrets clusters instance_profiles jobs metastore metastore_table_acls
```

Test:
Unit testing: `python3 -m unittest pipeline/pipeline_test.py`

Manual testing:
```
# skip everything
❯ python3 migration_pipeline.py --profile staging_src --export-pipeline --use-checkpoint --cluster-name kevinTestExport --skip-tasks instance_profiles users groups workspace_item_log workspace_acls notebooks secrets clusters instance_profiles jobs metastore metastore_table_acls instance_pools

Using the session id: 20211206231044
2021-12-06,23:10:44;INFO;Start export_instance_profiles
2021-12-06,23:10:44;INFO;export_instance_profiles Completed. Total time taken: 0:00:00.000155
2021-12-06,23:10:44;INFO;Start export_users
2021-12-06,23:10:44;INFO;export_users Completed. Total time taken: 0:00:00.000037
2021-12-06,23:10:44;INFO;Start export_groups
2021-12-06,23:10:44;INFO;export_groups Completed. Total time taken: 0:00:00.000031
2021-12-06,23:10:44;INFO;Start export_workspace_items_log
2021-12-06,23:10:44;INFO;export_workspace_items_log Completed. Total time taken: 0:00:00.000029
2021-12-06,23:10:44;INFO;Start export_workspace_acls
2021-12-06,23:10:44;INFO;export_workspace_acls Completed. Total time taken: 0:00:00.000028
2021-12-06,23:10:44;INFO;Start export_notebooks
2021-12-06,23:10:44;INFO;export_notebooks Completed. Total time taken: 0:00:00.000027
2021-12-06,23:10:44;INFO;Start export_secrets
2021-12-06,23:10:44;INFO;export_secrets Completed. Total time taken: 0:00:00.000027
2021-12-06,23:10:44;INFO;Start export_clusters
2021-12-06,23:10:44;INFO;export_clusters Completed. Total time taken: 0:00:00.000027
2021-12-06,23:10:44;INFO;Start export_instance_pools
2021-12-06,23:10:44;INFO;export_instance_pools Completed. Total time taken: 0:00:00.000043
2021-12-06,23:10:44;INFO;Start export_jobs
2021-12-06,23:10:44;INFO;export_jobs Completed. Total time taken: 0:00:00.000031
2021-12-06,23:10:44;INFO;Start export_metastore
2021-12-06,23:10:44;INFO;export_metastore Completed. Total time taken: 0:00:00.000029
2021-12-06,23:10:44;INFO;Start export_metastore_table_acls
2021-12-06,23:10:44;INFO;export_metastore_table_acls Completed. Total time taken: 0:00:00.000028
2021-12-06,23:10:44;INFO;Start finish_export
Export finished successfully. Session Id: 20211206231044
2021-12-06,23:10:44;INFO;finish_export Completed. Total time taken: 0:00:00.000101


# skip everything but users
❯ python3 migration_pipeline.py --profile staging_src --export-pipeline --use-checkpoint --cluster-name kevinTestExport --skip-tasks instance_profiles groups workspace_item_log workspace_acls notebooks secrets clusters instance_profiles jobs metastore metastore_table_acls instance_pools

Using the session id: 20211206231127
2021-12-06,23:11:27;INFO;Start export_instance_profiles
....
Get: https://dbc-dcc9b651-86e5.staging.cloud.databricks.com/api/2.0/preview/scim/v2/Users
....
2021-12-06,23:11:27;INFO;Start finish_export
Export finished successfully. Session Id: 20211206231127
2021-12-06,23:11:27;INFO;finish_export Completed. Total time taken: 0:00:00.000100

```

e2e testing without any skip
```
python3 migration_pipeline.py --profile prod_src_miklos --export-pipeline --use-checkpoint --cluster-name testing
python3 migration_pipeline.py --profile prod_dst_miklos --export-pipeline --use-checkpoint --cluster-name testing
```

### CODE REVIEW

File: migration_pipeline.py [hunk 1]
```diff
@@ -46,6 +49,24 @@ def build_pipeline(args) -> Pipeline:
     # TODO: Add verification job at the end
 
 
+def _add_task_or_skip(pipeline, skip_tasks, task, dependent_tasks):
```
Comment (yubing) [inline migration_pipeline.py:52]: I think it's better to add the skipping feature into pipeline framework. Similar to the dry_run. In that way, we can skipping tasks without changing pipeline structure.

Comment (kevinkim9264) [reply migration_pipeline.py:52]: Great idea. let me do that

Comment (kevinkim9264) [reply migration_pipeline.py]: done

**Review by yubing — APPROVED** [review]

File: pipeline/pipeline.py [hunk 2]
```diff
@@ -67,7 +67,7 @@ def _run_task(self, task: AbstractTask):
             return
         start = timer()
         logging.info(f'Start {task.name}')
-        if not self._dry_run:
+        if not self._dry_run and not task.skip:
```
Comment (kevinkim9264) [inline pipeline/pipeline.py:70]: This is what makes the pipeline not execute the task if its skip==true

File: dbclient/TableACLsClient.py [hunk 3]
```diff
@@ -231,8 +231,10 @@ def export_table_acls(self, db_name='', table_alcs_dir='table_acls/'):
 
         user_name = self.get_current_username(must_be_admin=True)
         if self.DB_ADMIN_SUFFIX in user_name:
-            user_name = "admin"
-        export_table_acls_workspace_path = f"/Users/{user_name}/tmp/migrate/Export_Table_ACLs.py"
+            notebook_parent_path = ""
+        else:
+            notebook_parent_path = f"/Users/{user_name}"
+        export_table_acls_workspace_path = f"{notebook_parent_path}/tmp/migrate/Export_Table_ACLs.py"
```
Comment (souravkhandelwal09) [inline dbclient/TableACLsClient.py:237]: why is this change showing up here ?

**Review by souravkhandelwal09 — APPROVED** [review]

File: dbclient/TableACLsClient.py (hunk 3 above)
Comment (kevinkim9264) [reply dbclient/TableACLsClient.py:237]: oh.. I am not sure. hopefully merging this will resolve it

**Pull request merged**